designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+handling are now divided between two languages; furthermore, this
+necessitates code be written to serialize/deserialize intermediates.
+
+A Rust interface, in similar spirit to the CmdStan interfaces from
+other languages, would provide the necessary abstraction to eliminate
+the aforementioned problems.
+
+# Functional Specification
+
+Given a Stan program, a user of the library will compile the model (if
+desired), call the executable with arguments (translated from
+strongly-typed argument tree), and obtain a self-contained context
+which encapsulates the pertinent information from the call.
+
+### Assumptions
+
+We assume (at our peril) that Rust programmers that will be able to
+figure out how to satisfy the following requirement:
+- a working CmdStan installation exists at some user-accessible path
+
+### Processes, IO
+
+The proposal is to use the Rust `std` library, in particular the
+[process](https://doc.rust-lang.org/std/process/index.html),
+[path](https://doc.rust-lang.org/std/path/index.html),
+[fs](https://doc.rust-lang.org/std/fs/index.html), and
+[ffi](https://doc.rust-lang.org/std/ffi/index.html) modules, to
+orchestrate processes, interact with file system and handle
+cross-platform concerns. This will yield a library which is portable,
+provided that it is (cross-)compiled for the intended target.
+
+## Control: compilation and calling the resultant executable
+
+A `CmdStanModel` type will serve as an abstraction for a Stan program,
+which may need to be compiled. Rather than compile on construction, a
+compilation method must be explicitly called in user code (assuming
+that a satisfactory executable does not exist yet).
+
+Two arguments will be necessary to create a `CmdStanModel`:
+1. a path to a CmdStan installation
+2. a path to a Stan program
+
+Methods (receiver: `CmdStanModel` instance) exposed to the user will
+include:
+- `validate_cmdstan` : determine whether the CmdStan installation works
+- `executable_works` : is there a working executable at the path
+  implied by the Stan file?
+- `compile_with_args` : attempt to compile a Stan program with
+  optional `make` arguments
+- `call_executable` : call executable with the given argument tree; on
+  success, return a `CmdStanOutput` instance.
+
+## Output
+
+Output of a successful `call_executable` call on a `CmdStanModel` will
+produce a `CmdStanOuput` instance, which encapsulates the context of
+the call. This includes:
+- the console output (exit status, stdout and stderr),
+- the argument tree provided to `call_executable`
+- the current working directory of the process at the time the call was made
+- the CmdStan installation from the parent `CmdStanModel`
+
+The objective is for `CmdStanOutput` to be a self-contained record
+which includes all pertinent information from a successful executable
+call. This structure can then be used to direct calls of
+`diagnose`/`stansummary`. Naturally, methods on said type will be
```

Comment:
Can you supply a complete list of signatures for methods on `CmdStanOutput`? You mention this path helper (very nice to provide, somewhat tricky to get right!) and `stansummary`, but what would their signatures look like, and are there more? I'm sure some of this will need to be discovered during implementation